### PR TITLE
Feature/arith corner cases test

### PR DESCRIPTION
**Pull request by fractasy**

### CODE REVIEW

File: integration-tests/tests/panic_emulator.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+fn execute_with_input(client: &EmbeddedExecuteOnlyClient, input: u64) -> anyhow::Result<()> {
+    let stdin = ZiskStdin::new();
+    stdin.write(&input);
+    client.execute(&ELF_PANIC_MODES, stdin, None)?;
+    Ok(())
+}
+
+#[test]
+#[ignore]
+fn panic_modes_emulator() {
```
[the remark this answers is not in the document]
Comment (fractasy) [reply integration-tests/tests/panic_emulator.rs:25]: It does not apply, after changing destination branch.

File: ziskos/entrypoint/src/hints/mod.rs [hunk 2]
```diff
@@ -32,7 +32,7 @@ use std::{
     sync::Arc,
 };
 use tokio::sync::oneshot;
-use zisk_common::io::{StreamWrite, UnixSocketStreamWriter};
+use zisk_stream::{StreamWrite, UnixSocketStreamWriter};
```
[the remark this answers is not in the document]
Comment (fractasy) [reply ziskos/entrypoint/src/hints/mod.rs:36]: It does not apply, after changing destination branch.

File: integration-tests/tests/panic_assembly.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+fn execute_with_input(client: &EmbeddedExecuteOnlyClient, input: u64) -> anyhow::Result<()> {
+    let stdin = ZiskStdin::new();
+    stdin.write(&input);
+    client.execute(&ELF_PANIC_MODES, stdin, None)?;
+    Ok(())
+}
+
+#[test]
+#[ignore]
+fn panic_modes_assembly() {
```
[the remark this answers is not in the document]
Comment (fractasy) [reply integration-tests/tests/panic_assembly.rs:29]: It does not apply, after changing destination branch.

File: core/src/ops_core.rs [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+pub const fn op_sll(a: u64, b: u64) -> (u64, bool) {
+    (a << (b & 0x3f), false)
+}
+
+/// Shifts a as a 32-bits unsigned value to the left b mod 64 bits, and stores the result in c (and
+/// sets flag to false)
+#[inline(always)]
+pub fn op_sll_w(a: u64, b: u64) -> (u64, bool) {
+    (((Wrapping(a as u32) << (b & 0x3f) as usize).0 as i32) as u64, false)
+}
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs]: Fixed.

File: core/src/ops_core.rs [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+pub const fn op_srl(a: u64, b: u64) -> (u64, bool) {
+    (a >> (b & 0x3f), false)
+}
+
+/// Shifts a as a 32-bits signed value to the right b mod 64 bits, and stores the result in c (and
+/// sets flag to false)
+#[inline(always)]
+pub fn op_sra_w(a: u64, b: u64) -> (u64, bool) {
+    ((Wrapping(a as i32) >> (b & 0x3f) as usize).0 as u64, false)
+}
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs]: Fixed.

File: core/src/ops_core.rs [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+
+/// Shifts a as a 32-bits unsigned value to the right b mod 64 bits, and stores the result in c (and
+/// sets flag to false)
+#[inline(always)]
+pub fn op_srl_w(a: u64, b: u64) -> (u64, bool) {
+    (((Wrapping(a as u32) >> (b & 0x3f) as usize).0 as i32) as u64, false)
+}
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs]: Fixed.

File: core/src/ops_core.rs [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+/// Sets c to a / b, as 32-bits signed values, and flag to false.
+/// If b=0 (divide by zero) it sets c to 2^64 - 1, and sets flag to true.
+#[inline(always)]
+pub const fn op_div_w(a: u64, b: u64) -> (u64, bool) {
+    if b as i32 == 0 {
+        return (M64, true);
+    }
+
+    ((((a as i32) as i64) / ((b as i32) as i64)) as u64, false)
+}
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs:350]: Fixed.

File: test-artifacts/programs/diagnostic/src/riscv_ima.rs [hunk 8]
```diff
@@ -173,9 +177,83 @@ pub fn diagnostic_riscv_ima() {
...
+            mul_w(a as i32, b as i32, (op_mul_w(a, b).0 & 0xFFFF_FFFF) as i32);
+
+            // Avoid divide by zero (b = 0)
+            // Avoid integer overflow (a = MIN_N_32, b = -1)
+            if b != 0 && (a != MIN_N_32 || b != 0xFFFF_FFFF/* -1 */) {
+                div_w(a as i32, b as i32, (op_div_w(a, b).0 & 0xFFFF_FFFF) as i32);
+                divu_w(a as u32, b as u32, (op_divu_w(a, b).0 & 0xFFFF_FFFF) as u32);
+                rem_w(a as i32, b as i32, (op_rem_w(a, b).0 & 0xFFFF_FFFF) as i32);
+                remu_w(a as u32, b as u32, (op_remu_w(a, b).0 & 0xFFFF_FFFF) as u32);
+            }
```
[the remark this answers is not in the document]
Comment (fractasy) [reply test-artifacts/programs/diagnostic/src/riscv_ima.rs]: Fixed.  All corner cases are executed now, even divide by zero.

File: core/src/ops_core.rs [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+pub const fn op_rem_w(a: u64, b: u64) -> (u64, bool) {
+    // Handle divide by zero case
+    if (b as i32) == 0 {
+        return ((a as i32) as u64, true);
+    }
+
+    // Handle overflow case: -MIN_I32 cannot be represented in 32 bits, so return a.
+    if a as u32 == 0x8000_0000 && b as i32 == -1 {
+        return (0, true);
+    }
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 10]
```diff
@@ -4656,7 +4656,7 @@ impl ZiskRom2Asm {
                     REG_C,
                     ctx.comment_str("Div: set result to f's")
                 );
-
+                *code += &format!("\tmov {}, 1 {}\n", REG_FLAG, ctx.comment_str("flag = 1"));
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 11]
```diff
@@ -4927,36 +4939,70 @@ impl ZiskRom2Asm {
...
                     REG_C,
                     ctx.comment_str("DivW: set result to f's")
                 );
+                *code += &format!("\tmov {}, 1 {}\n", REG_FLAG, ctx.comment_str("flag = 1"));
+                *code += &format!("\tjmp pc_{:x}_divw_done\n", ctx.pc);
 
-                *code += &format!("\tje pc_{:x}_divw_done\n", ctx.pc);
+                // Check overflow:
+                // If a=0x80000000 (MIN_I32) and b=0xffffffff (-1) the result should be -MIN_I32,
+                // which cannot be represented with 32 bits, so return 0xffffffffffffffff
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 12]
```diff
@@ -4927,36 +4939,70 @@ impl ZiskRom2Asm {
...
+                );
+                *code += &format!(
+                    "\tjne pc_{:x}_divw_divide {}\n",
+                    ctx.pc,
+                    ctx.comment_str("DivW: if b is not -1, divide")
+                );
+                *code += &format!(
+                    "\tmov {}, 0xffffffff80000000 {}\n",
+                    REG_C,
+                    ctx.comment_str("DivW: set result to f's")
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs]: Fixed.

File: core/src/ops_core.rs [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+
+    (a / b, false)
+}
+
+/// Sets c to a / b, as 64-bits signed values, and flag to false.
+///
+/// If b=0 (divide by zero) it sets c to 2^64 - 1, and sets flag to true.
+/// If a=0x8000000000000000 (MIN_I64) and b=0xFFFFFFFFFFFFFFFF (-1) the result should be -MIN_I64,
+/// which cannot be represented with 64 bits (overflow) and it returns c=a.
+#[inline(always)]
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs:306]: Fixed.

File: core/src/ops_core.rs [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+pub const fn op_rem(a: u64, b: u64) -> (u64, bool) {
+    // Handle divide by zero case
+    if b == 0 {
+        return (a, true);
+    }
+
+    // Handle overflow case: -MIN_I64 cannot be represented in 64 bits, so return a.
+    if a as u64 == 0x8000_0000_0000_0000 && b as i64 == -1 {
+        return (0, true);
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs]: Fixed.

File: core/src/ops_core.rs [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+    if b as u32 == 0 {
+        return (M64, true);
+    }
+
+    (((a as u32 / b as u32) as i32) as u64, false)
+}
+
+/// Sets c to a / b, as 32-bits signed values, and flag to false.
+/// If b=0 (divide by zero) it sets c to 2^64 - 1, and sets flag to true.
+#[inline(always)]
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs:337]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 16]
```diff
@@ -4979,37 +5025,72 @@ impl ZiskRom2Asm {
...
                 // Unsigned divide RDX:RAX by r/m64, with result stored in RAX := Quotient, RDX :=
                 // Remainder.
 
                 // Check divide by zero:
-                // If b==0 return a
+                // If b=0 (divide by zero) it sets c to 0, and sets flag to true.
                 *code += &format!(
                     "\tcmp {}, 0 {}\n",
                     REG_B_W,
                     ctx.comment_str("RemW: if b == 0 return f's")
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs]: Fixed.

File: core/src/ops_core.rs [hunk 17]
```diff
@@ -0,0 +1,489 @@
...
+/// If b=0 (divide by zero) it sets c to 2^64 - 1, and sets flag to true.
+/// If a=0x80000000 (MIN_I32) and b=0xFFFFFFFF (-1) it returns 0xffffffff80000000 and sets flag to true.
+#[inline(always)]
+pub const fn op_div_w(a: u64, b: u64) -> (u64, bool) {
+    // Handle divide by zero case
+    if b as i32 == 0 {
+        return (M64, true);
+    }
+
+    // Handle overflow case: -MIN_I32 cannot be represented in 32 bits, so return a.
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/ops_core.rs]: Fixed.

File: core/src/zisk_ops.rs [hunk 18]
```diff
@@ -31,6 +31,8 @@ use std::{
 use tiny_keccak::keccakf;
 use ziskos::zisklib::FCALL_INPUT_READY_ID;
 
+use crate::ops_core::*;
+use crate::ops_core_context::*;
 use crate::{FCALL_PARAMS_MAX_SIZE, FCALL_RESULT_MAX_SIZE};
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_ops.rs:36]: I prefer to explicitly use ops_core and ops_context

File: core/src/zisk_rom_2_asm.rs [hunk 19]
```diff
@@ -4699,7 +4702,7 @@ impl ZiskRom2Asm {
                     REG_A,
                     ctx.comment_str("Div: set result to a")
                 );
-
+                *code += &format!("\tmov {}, 1 {}\n", REG_FLAG, ctx.comment_str("flag = 1"));
                 *code += &format!("\tje pc_{:x}_div_done\n", ctx.pc);
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 20]
```diff
@@ -4733,9 +4736,10 @@ impl ZiskRom2Asm {
                     REG_C,
                     ctx.comment_str("Div: c = quotient(rax)")
                 );
+                *code += &format!("\tmov {}, 0 {}\n", REG_FLAG, ctx.comment_str("flag = 0"));
+
                 *code += &format!("pc_{:x}_div_done:\n", ctx.pc);
                 ctx.c.is_saved = true;
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs:4767]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 21]
```diff
@@ -4766,8 +4783,8 @@ impl ZiskRom2Asm {
                     REG_A,
                     ctx.comment_str("Rem: set result to a")
                 );
-
-                *code += &format!("\tje pc_{:x}_rem_done\n", ctx.pc);
+                *code += &format!("\tmov {}, 1 {}\n", REG_FLAG, ctx.comment_str("flag = 1"));
+                *code += &format!("\tjmp pc_{:x}_rem_done\n", ctx.pc);
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs:4796]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 22]
```diff
@@ -4733,9 +4748,11 @@ impl ZiskRom2Asm {
...
                 );
+                *code += &format!("\tmov {}, 0 {}\n", REG_FLAG, ctx.comment_str("flag = 0"));
+
+                // Done
                 *code += &format!("pc_{:x}_div_done:\n", ctx.pc);
                 ctx.c.is_saved = true;
-                ctx.flag_is_always_zero = true;
             }
             ZiskOp::Rem => {
                 assert!(ctx.store_a_in_a);
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs:4770]: Fixed.

**Review by agnusmor — DISMISSED** [review]

File: core/src/zisk_rom_2_asm.rs [hunk 23]
```diff
@@ -4810,14 +4827,9 @@ impl ZiskRom2Asm {
                     ctx.pc,
                     ctx.comment_str("Rem: if b is not 0xffffffffffffffff, divide")
                 );
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs:4832]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 24]
```diff
@@ -4755,7 +4772,7 @@ impl ZiskRom2Asm {
                 // Remainder
 
                 // Check divide by zero:
-                // If b==0 return 0xffffffffffffffff
+                // If b==0 return a
                 *code += &format!(
                     "\tcmp {}, 0 {}\n",
                     REG_B,
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 25]
```diff
@@ -4739,9 +4754,11 @@ impl ZiskRom2Asm {
...
                 );
+                *code += &format!("\tmov {}, 0 {}\n", REG_FLAG, ctx.comment_str("flag = 0"));
+
+                // Done
                 *code += &format!("pc_{:x}_div_done:\n", ctx.pc);
                 ctx.c.is_saved = true;
-                ctx.flag_is_always_zero = true;
             }
             ZiskOp::Rem => {
                 assert!(ctx.store_a_in_a);
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs:4770]: Fixed.

File: test-artifacts/programs/diagnostic/src/riscv_ima.rs [hunk 26]
```diff
@@ -1,3 +1,7 @@
+#[path = "../../../../core/src/ops_core.rs"]
+mod ops_core;
+
+use ops_core::*;
```
[the remark this answers is not in the document]
Comment (fractasy) [reply test-artifacts/programs/diagnostic/src/riscv_ima.rs:7]: Will not fix.

File: test-artifacts/programs/diagnostic/src/riscv_ima.rs [hunk 27]
```diff
@@ -1,3 +1,7 @@
...
 use std::arch::asm;
```
[the remark this answers is not in the document]
Comment (fractasy) [reply test-artifacts/programs/diagnostic/src/riscv_ima.rs:8]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 28]
```diff
@@ -4538,14 +4540,16 @@ impl ZiskRom2Asm {
...
                 // Unsigned divide RDX:RAX by r/m64, with result stored in RAX := Quotient, RDX :=
                 // Remainder
+
+                // Divide by zero:
                 // If b==0 return 0xffffffffffffffff
                 *code += &format!(
                     "\tcmp {}, 0 {}\n",
                     REG_B,
                     ctx.comment_str("Divu: if b == 0 return f's")
                 );
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs:4556]: Fixed.

File: core/src/zisk_rom_2_asm.rs [hunk 29]
```diff
@@ -4579,34 +4585,40 @@ impl ZiskRom2Asm {
...
                 // Unsigned divide RDX:RAX by r/m64, with result stored in RAX := Quotient, RDX :=
                 // Remainder
+
+                // Divide by zero:
                 // If b==0 return a
                 *code += &format!(
                     "\tcmp {}, 0 {}\n",
                     REG_B,
                     ctx.comment_str("Remu: if b == 0 return a")
                 );
```
[the remark this answers is not in the document]
Comment (fractasy) [reply core/src/zisk_rom_2_asm.rs:4611]: Fixed.

**Review by zkronos73 — APPROVED** [review]
LGTM
